python-cfd/CG.py

```diff
@@ -0,0 +1,66 @@
+import sys
+import math
+import numpy as np
+import scipy as sp
+from scipy import linalg
+
+A = np.array([[-2,1,1,0],[1,-2,0,1],[1,0,-2,1],[0,1,1,-2]])
+b = np.array([1, -1, 1, -1])
+
+# Provide some information
+print("# matrix:")
+for r in A:
+    print("   ", r)
+print("  eigenvalues:", ", ".join(["{:.6g}".format(e) for e in np.linalg.eigvals(A)]))
+print("  rank:", np.linalg.matrix_rank(A))
+
+kerA = sp.linalg.null_space(A)
+print("  nullspace:")
+for r in kerA:
+    print("   ", r)
+
+# Initialize containers and values
+print("# intialization:")
+x = np.array([0., 0., 0., 0.])
+print("  initial guess:", x)
+
+residuals = [b - A @ x]
+rms2 = residuals[0] @ residuals[0]
+print("  residual:{} (L2 norm: {})".format(residuals[0], rms2))
+
+directions = residuals[:]
+print("  directions", [list(d) for d in directions])
+alpha, beta = [], []
```

Comment:
why are we making a list for alpha and beta? we don't really need to track the history

Reply:
Agreed, in C++ implementation we don't but I kept them in this demonstrator just to verify/test the numerical performance of the algo.